### PR TITLE
Make GetLogEntryDetails return 404 if the log entry doesn't exist ins…

### LINKED ISSUE #1618: Should return 404 or 400 if the deployment log doesn't exist instead of returning a 500
In `api/deployments/{deploymentId}/log/{logId}`
If `{logId}` doesn't exist, we throw an unhandled exception and return a 500.

``` javascript
{
    Message: "An error has occurred.",
    ExceptionMessage: "Sequence contains no elements",
    ExceptionType: "System.InvalidOperationException",
    StackTrace: " at System.Linq.Enumerable.First[TSource](IEnumerable`1 source) at Kudu.Core.Deployment.XmlLogger.GetLogEntryDetails(String entryId) at Kudu.Core.Deployment.DeploymentManager.GetLogEntryDetails(String id, String entryId) at Kudu.Services.Deployment.DeploymentController.GetLogEntryDetails(String id, String logId) at lambda_method(Closure , Object , Object[] ) at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ActionExecutor.<>c__DisplayClass10.<GetExecutor>b__9(Object instance, Object[] methodParameters) at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ActionExecutor.Execute(Object instance, Object[] arguments) at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ExecuteAsync(HttpControllerContext controllerContext, IDictionary`2 arguments, CancellationToken cancellationToken) --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at System.Web.Http.Controllers.ApiControllerActionInvoker.<InvokeActionAsyncCore>d__0.MoveNext() --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__5.MoveNext() --- End of stack trace from previous location where exception was thrown --- at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__5.MoveNext() --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at System.Web.Http.Filters.ActionFilterAttribute.<ExecuteActionFilterAsyncCore>d__0.MoveNext() --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__5.MoveNext() --- End of stack trace from previous location where exception was thrown --- at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__5.MoveNext() --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at System.Web.Http.Filters.ActionFilterAttribute.<ExecuteActionFilterAsyncCore>d__0.MoveNext() --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at System.Web.Http.Controllers.ActionFilterResult.<ExecuteAsync>d__2.MoveNext() --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at System.Web.Http.Controllers.ExceptionFilterResult.<ExecuteAsync>d__0.MoveNext() --- End of stack trace from previous location where exception was thrown --- at System.Web.Http.Controllers.ExceptionFilterResult.<ExecuteAsync>d__0.MoveNext() --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at System.Web.Http.Dispatcher.HttpControllerDispatcher.<SendAsync>d__1.MoveNext()"
}
```

### PR DESCRIPTION
**Pull request by ahmelsayed**
…tead of throwing a 500, fixes #1618

### CODE REVIEW

File: Kudu.Core/Deployment/XmlLogger.cs [hunk 1]
```diff
@@ -62,10 +62,17 @@ public IEnumerable<LogEntry> GetLogEntryDetails(string entryId)
...
-                   let type = (LogEntryType)Int32.Parse(e.Attribute("type").Value)
-                   select new LogEntry(time, e.Attribute("id").Value, e.Element("message").Value, type);
+            return document.Root.Elements("entry")
+                .Where(s => s.Attribute("id").Value == entryId)
+                .Select(s => s.Elements("entry"))
+                .Select(s => s.Select(e =>
+                {
+                    var time = DateTime.Parse(e.Attribute("time").Value).ToUniversalTime();
+                    var type = (LogEntryType)Int32.Parse(e.Attribute("type").Value);
+                    return new LogEntry(time, e.Attribute("id").Value, e.Element("message").Value, type);
```
Comment (shrimpy) [inline Kudu.Core/Deployment/XmlLogger.cs]: would it be better if we use TryParse, just in-case log format change and we will end up 500 error again

Comment (ahmelsayed) [inline Kudu.Core/Deployment/XmlLogger.cs]: Wouldn't a 500 be appropriate in that case since it's a server error to have a wrong log?

Comment (shrimpy) [inline Kudu.Core/Deployment/XmlLogger.cs]: i just worry later if we change the log format and forget to update this piece of code. maybe have a test case to ensure the format, if log format change then at least testcase will tell us to update code here.

File: Kudu.Core/Deployment/XmlLogger.cs [hunk 2]
```diff
@@ -62,10 +62,17 @@ public IEnumerable<LogEntry> GetLogEntryDetails(string entryId)
...
+                }))
+                .FirstOrDefault() ?? Enumerable.Empty<LogEntry>();
```
Comment (mathewc) [inline Kudu.Core/Deployment/XmlLogger.cs]: Is returning an empty enumerable a new behavior? If so, please add test(s) to XmlLoggerFacts for it

File: Kudu.Services/Deployment/DeploymentController.cs [hunk 3]
```diff
@@ -210,7 +210,12 @@ public HttpResponseMessage GetLogEntryDetails(string id, string logId)
                 try
                 {
                     var details = _deploymentManager.GetLogEntryDetails(id, logId).ToList();
-                    return Request.CreateResponse(HttpStatusCode.OK, ArmUtils.AddEnvelopeOnArmRequest(details, Request));
+                    return details.Any()
```
Comment (mathewc) [inline Kudu.Services/Deployment/DeploymentController.cs]: Add a test to DeploymentControllerFacts for this

File: Kudu.Core/Deployment/XmlLogger.cs (hunk 2 above)
Comment (ahmelsayed) [inline Kudu.Core/Deployment/XmlLogger.cs]: Added [test cases](https://github.com/projectkudu/kudu/commit/2b1992c86c2d7c97a188c866ca0256fb5d32ecba)

File: Kudu.Services/Deployment/DeploymentController.cs (hunk 3 above)
Comment (ahmelsayed) [inline Kudu.Services/Deployment/DeploymentController.cs]: https://github.com/projectkudu/kudu/blob/2b1992c86c2d7c97a188c866ca0256fb5d32ecba/Kudu.Services.Test/DeploymentControllerFacts.cs#L66-L83

**Comment by mathewc** [comment]
Looks good to me, thanks for adding tests :shipit:

**Comment by shrimpy** [comment]
:shipit:

**Pull request merged**